### PR TITLE
Feature/implement hand manager

### LINKED ISSUE #17: Implement HandManager

### PR DESCRIPTION
**Pull request by meszszi**
Resolves #17 

I think we should focus only on the hand flow logic in this PR and not worry about the design of updates, game manager and communicators, as they were created mainly to allow for running the game in raw developer mode.

### CODE REVIEW

File: server/src/main/kotlin/core/gameflow/BettingRound.kt [hunk 1]
```diff
@@ -4,5 +4,15 @@ enum class BettingRound {
...
-    RIVER
+    RIVER;
+
+
+    fun next(): BettingRound {
+        return when(this) {
+            PRE_FLOP -> FLOP
+            FLOP -> TURN
+            TURN -> RIVER
+            else -> TODO()
```
Comment (meszszi) [inline server/src/main/kotlin/core/gameflow/BettingRound.kt]: add some proper exception

File: server/src/main/kotlin/core/gameflow/dealer/Dealer.kt [hunk 2]
```diff
@@ -1,14 +1,25 @@
...
+import core.gameflow.BettingRound
 import core.gameflow.handstate.HandState
 import core.gameflow.handstate.rebuild
 import kotlin.random.Random
 
-class Dealer {
+class Dealer : IDealer {
     private var deckIterator: ListIterator<Card>? = null
 
+    // TODO : Test
```
Comment (meszszi) [inline server/src/main/kotlin/core/gameflow/dealer/Dealer.kt]: add tests

File: server/src/test/kotlin/core/gameflow/player/PlayerTest.kt [hunk 3]
```diff
@@ -184,4 +184,21 @@ class PlayerTest {
         assert(playerHand is TwoPair)
         assert(playerHand.cards == properHandCards)
     }
+
+    @Test
+    fun `moveBetToPot should move a playrs bet to pot`() {
```
Comment (meszszi) [inline server/src/test/kotlin/core/gameflow/player/PlayerTest.kt]: typo in 'players'

File: server/src/main/kotlin/core/adapters/LocalPlayerAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            "RAISE" -> {
+                val size = words[1].toInt()
+                Raise(size)
+            }
+            "ALLIN" -> AllIn
+            else -> throw IllegalArgumentException("could not parse any action from command '$command'")
+        }
+    }
+}
```
Comment (meszszi) [inline server/src/main/kotlin/core/adapters/LocalPlayerAdapter.kt]: this is ugly af xd

File: server/src/main/kotlin/core/gameflow/gamestate/PlayerInfo.kt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package core.gameflow.gamestate
+
+data class PlayerInfo(val position: Int, val stack: Int) {
+    val id: Int = position
+}
```
Comment (meszszi) [inline server/src/main/kotlin/core/gameflow/gamestate/PlayerInfo.kt]: should it be moved to the same file as GameState (and maybe renamed)?

File: server/src/main/kotlin/core/gameflow/handstate/HandManager.kt [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        return initialStateBuilder.build()
+    }
+
+    private fun HandState.handleAction(): HandState {
+        var handState = this
+
+        while (true) {
+            handState = dealer.deal(handState)
+            communicator.broadcastUpdate(HandStateUpdate(handState))
+            handState = bettingRound(handState)
```
Comment (meszszi) [inline server/src/main/kotlin/core/gameflow/handstate/HandManager.kt]: it's not really obvious what's going on, maybe 'runBettingRound'?

File: server/src/main/kotlin/core/gameflow/handstate/HandManager.kt [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+import core.gameflow.*
+import core.gameflow.dealer.Dealer
+import core.gameflow.dealer.IDealer
+import core.gameflow.gamestate.GameState
+import core.gameflow.gamestate.PlayerInfo
+import core.gameflow.player.*
+
+class HandManager(val dealer: IDealer, val communicator: Communicator, val roomSettings: RoomSettings) : IHandManager {
+
+    override fun playHand(gameState: GameState, newPlayersIds: Set<Int>): GameState = gameState
```
Comment (meszszi) [inline server/src/main/kotlin/core/gameflow/handstate/HandManager.kt]: This method should be thoroughly tested

File: server/src/main/kotlin/core/gameflow/handstate/HandStateExtensions.kt [hunk 8]
```diff
@@ -7,8 +7,7 @@ import core.gameflow.player.Player
 import core.gameflow.Positions
 
 fun HandState.updateActivePlayer(playerUpdate: Player): HandState {
-
-    assert(activePlayer != null)
+    activePlayer!!
```
Comment (wujuu) [inline server/src/main/kotlin/core/gameflow/handstate/HandStateExtensions.kt]: how is it different than the deleted line? just asking xd

File: server/src/test/kotlin/core/gameflow/handstate/HandManagerTest.kt [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        assert(finalState.positions == Positions(0, 1, 2))
+        assert(finalState.playersInfo == listOf(
+                PlayerStatus(0, 1000),
+                PlayerStatus(1, 200),
+                PlayerStatus(2, 900),
+                PlayerStatus(3, 1900)
+        ))
+
+        // check that dealer has dealt hole cards and all three streets
+        verifySequence {
```
Comment (wujuu) [inline server/src/test/kotlin/core/gameflow/handstate/HandManagerTest.kt]: what does this function do?

File: server/src/main/kotlin/core/gameflow/handstate/HandStateExtensions.kt (hunk 8 above)
Comment (meszszi) [reply server/src/main/kotlin/core/gameflow/handstate/HandStateExtensions.kt]: It is not, it just looks a bit nicer imo

File: server/src/test/kotlin/core/gameflow/handstate/HandManagerTest.kt (hunk 9 above)
Comment (meszszi) [reply server/src/test/kotlin/core/gameflow/handstate/HandManagerTest.kt]: read the docs

**Pull request merged**